// ngrx-vis/resolvers/store-dispatcher.resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = require("path");
var ts_morph_1 = require("../utils/ts-morph");
var StoreDispatcherResolver = /** @class */ (function () {
    function StoreDispatcherResolver() {
    }
    StoreDispatcherResolver.prototype.canExecute = function (actionReference) {
        var actionCreatorCall = ts_morph_1.getCaller(actionReference.getNode());
        var actionDispatchCall = actionCreatorCall
            ? ts_morph_1.getCaller(actionCreatorCall)
            : null;
        return !actionDispatchCall
            ? false
            : actionDispatchCall.getText().includes('.dispatch(');
    };
    StoreDispatcherResolver.prototype.execute = function (actionReference) {
        return {
            fileName: path_1.basename(actionReference.getSourceFile().getFilePath()),
            filePath: actionReference.getSourceFile().getFilePath(),
            category: 'component'
        };
    };
    return StoreDispatcherResolver;
}());
exports.StoreDispatcherResolver = StoreDispatcherResolver;
